R/age_model_class.R

```diff
@@ -0,0 +1,222 @@
+#'
+#' Default time series is seq(0, 100, by = 1).
+#' This function relies on the packages deSolve and ggplot2.
+#' This function creates a plot of the variables over time and returns a
+#' vectors of the incidence numbers for each age group.
+#'
+#' @param object An object of the class age_model.
+#' @param times (vector) time sequence over which to solve the model.
+#'        Must be of the form seq(t_start,t_end,by=t_step).
+#' @param is_plot (logical) whether plots for the compartments for the
+#'        different age groups are drawn.
+#'
+#' @return data frame of time and outputs with incidence numbers for each
+#'         age group.
+setMethod(
+  'simulate', 'age_model',
+  function(object, times, is_plot) {
+    if(!is.double(times)){
+      stop('Evaluation times of the model storage format must be a vector.')
+    }
+    if(!is.logical(is_plot)){
+      stop('Trigger for plotting must be logical.')
+    }
+    age <- object@n_age_categories
+
+    #set initial state vector
+    state <- c(S = get_parameters(object)$S0,
+               E = get_parameters(object)$E0,
+               I = get_parameters(object)$I0,
+               R = get_parameters(object)$R0)
+
+    #set parameters vector
+    parameters <- c(a = get_parameters(object)$a,
+                    b = get_parameters(object)$b,
+                    c = get_parameters(object)$c)
+
+    #function for RHS of ode system
+    right_hand_side <- function(t, state, parameters) {
+      with(
+        as.list(c(state, parameters)),
+        {
+          S <- state[1:age]
+          E <- state[(age+1):(2*age)]
+          I <- state[(2*age+1):(3*age)]
+          R <- state[(3*age+1):(4*age)]
+          # rate of change
+          dS <- -a*S*I
+          dE <- a*S*I - b*E
+          dI <- b*E - c*I
+          dR <- c*I
+          # return the rate of change
+          list(c(dS, dE, dI, dR))
+        })
+    }
+
+    #call ode solver
+    out <- ode(
+      y = state, times = times, func = right_hand_side, parms = parameters)
+
+    output <- as.data.frame.array(out)
+
+    if(is_plot == TRUE){
+      col <- c('cS' = 'blue', 'cE' = 'green', 'cI' = 'yellow', 'cR' = 'red')
+      for(a in 1:age){
+        out_df <- output[,c(1, a+1, age+a+1, 2*age+a+1, 3*age+a+1)]
+        names(out_df) <- c('time', 'S', 'E', 'I', 'R')
+        SEIRplot <- ggplot(out_df, aes(x=time)) +
+          geom_line(aes(y=S, color = 'cS'), size = 1.5) +
+          geom_line(aes(y=E, color = 'cE'), size = 1.5) +
+          geom_line(aes(y=I, color = 'cI'), size = 1.5) +
+          geom_line(aes(y=R, color = 'cR'), size = 1.5) +
+          labs(x = 'time', y = 'fraction of the population',
+               title = glue('SEIR model for age group {a}')) +
+          theme(legend.position = 'right') +
+          scale_color_manual(values = col)
+        #show plot
+        print(SEIRplot)
+
+        #printtype of the plotting object to the command line for testing
+        print(typeof(SEIRplot))
+      }
+    }
```

Comment:
I know this was briefly discussed in the meeting so maybe I am beating a dead horse but I am not quite convinced by the need for a plotting option within the simulate function. I get that we want a quick option to verify results but if we eventually want to write a separate function for "nice" plots that handles the output from any function, I think it would be just as easy to add a flag there that let's you switch between a basic and a more detailed plot. Plotting of any kind would then be consistent between models and we don't need to have this chunk of plotting code in every single class separately. I will copy this into our basic SEIR class in case we want to stick with it but I think it's something to consider :)

Reply:
Having thought about this, I think I agree with @solveigvandervegt. I think it'll be easier to make generic plotting methods that'll work between classes of SEIR model if we keep them separate from the classes themselves. Otherwise, you're likely to have, what is termed as a "inappropriate intimacy" code smell: https://refactoring.guru/smells/inappropriate-intimacy

Reply:
Thanks for raising this both. I see how it might be more trouble than it's worth to having plotting routines embedded into the definition of the class itself given that the user will most likely want to have more control over any plotting they do. I've removed the routine for now (i'll keep it somewhere locally or on slack in case we need it later).